## 3. Splitting to test and train data, ensuring balance of classes

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pickle
fp = "/content/drive/My Drive/BT4222 Group Project/Final Project/Codes/Jar of Pickles/twitter_cleaned.pkl"
with open(fp,"rb") as f:
    twitter_cleaned = pickle.load(file=f)

twitter_cleaned.head()

tweets_cleaned  neutral  hate
0            woman complain cleaning house man trash        1     0
1   boy dats cold  tyga dwn bad cuffin dat hoe st...        0     1
2           dawg  fuck bitch start cry confused shit        0     1
3                                   look like tranny        0     1
4                shit hear true faker bitch told ya         0     1

In [3]:
for i,j in enumerate(twitter_cleaned['neutral'].value_counts().sort_index()):
    print(i,j,j/twitter_cleaned.shape[0]*100)

0 20620 83.20219505306056
1 4163 16.797804946939436


Here the class labels are as such and are imbalanced:

0 = Hate (83.20%)
<br/> 1 = Neutral (16.80%)<br>


For the feature-target split:

Feature: tweet data
Target: class, in the form of [x, x] representing [neutral, hate]

In [4]:
import numpy as np
def split_input_output(data):
    '''
    Function returns input and output variables as separate arrays
    '''
    
    class_labels = ["neutral","hate"]
    
    # separating input and output 
    x = list(data["tweets_cleaned"])
    
    
    y = np.array(data[class_labels])
    return x,y

### Selecting the tweet as the feature, and class array as target
X, y = split_input_output(data=twitter_cleaned)

In [5]:
### To check in feature-target split
print(X[:5])
print(y[:5])

[' woman complain cleaning house man trash', ' boy dats cold  tyga dwn bad cuffin dat hoe st place', ' dawg  fuck bitch start cry confused shit', ' look like tranny', ' shit hear true faker bitch told ya ']
[[1 0]
 [0 1]
 [0 1]
 [0 1]
 [0 1]]


In [6]:
from sklearn.model_selection import train_test_split

# Split into train + validation set & test set
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size = 0.1, 
                                                            shuffle=True, stratify=y, random_state = 42)

print(len(X_train_val))
print(y_train_val.shape)
print(len(X_test))
print(y_test.shape)

22304
(22304, 2)
2479
(2479, 2)


In [7]:
# Split into train & validation set
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size = 0.2, 
                                                  shuffle = True, stratify = y_train_val, random_state = 42)

print(len(X_train))
print(y_train.shape)
print(len(X_val))
print(y_val.shape)

17843
(17843, 2)
4461
(4461, 2)


In [8]:
# saving train, cv and test processed data and labels
fp = "/content/drive/My Drive/BT4222 Group Project/Final Project/Codes/Jar of Pickles/train_val_test_data.pkl"
with open(fp,mode="wb") as f:
    pickle.dump(obj=(X_train,
                     y_train,
                     X_val,
                     y_val,
                     X_test,
                     y_test),
                file=f)